primitives/src/merkle_tree/namespaced_merkle_tree/mod.rs

```diff
@@ -421,14 +451,40 @@ mod nmt_tests {
         assert!(Hasher::digest(&hashes).is_err());
     }
 
+    enum BuildType {
+        FromElems,
+        Push,
+        Extend,
+    }
+
     #[test]
     fn test_nmt() {
+        test_nmt_with_build_type(BuildType::Extend);
+        test_nmt_with_build_type(BuildType::FromElems);
+        test_nmt_with_build_type(BuildType::Push);
+    }
+
+    fn test_nmt_with_build_type(build_type: BuildType) {
         let namespaces = [1, 2, 2, 2, 4, 4, 4, 5];
         let first_ns = namespaces[0];
         let last_ns = namespaces[namespaces.len() - 1];
         let internal_ns = namespaces[1];
         let leaves: Vec<Leaf> = namespaces.iter().map(|i| Leaf::new(*i)).collect();
-        let tree = TestNMT::from_elems(3, leaves.clone()).unwrap();
+        let tree = match build_type {
```

Comment:
Do we expect the trees to be the same regardless of how they are built?

If this is the case maybe a dedicated test to check that would be worth it.  Because the current test can pass with different trees, for example if we replace lines 476 and 477 by

```rust
     let mut nmt = TestNMT::from_elems(3, &[leaves[0],leaves[1]]).unwrap();
     nmt.extend([leaves[2],leaves[3],leaves[4],leaves[5]]).unwrap(); // No elements leaves[6], leaves[7]
```
 the test still passes.

Reply:
yes I believe that trees should be same as long as leaves are inserted in the same order @mrain is this a fair assumption? If that is the case, I can write a dedicated test that compares the roots and the namespace ranges for consistency